src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,275 @@
+    if (opt_container) {
+      opt_container =
+          moveLayoutRect(opt_container, viewport.left, viewport.top);
+    }
+    let elementRectNorm;
+
+    if (this.isAmpElement_) {
+      // If calls with container, always assume getLayoutBox() return relative
+      // LayoutRect to container
+      const elementRect = this.element_.getLayoutBox();
+      elementRectNorm = opt_container
+          ? moveLayoutRect(elementRect, opt_container.left, opt_container.top)
+          : elementRect;
+    } else {
+      // If the element is not an AMP element, always assume its position to
+      // the page won't change.
+      if (!this.elementRectForNonAMP_) {
+        const elementRect = LayoutRectFromDomRect(
+            this.element_./*OK*/getBoundingClientRect());
+        this.elementRectForNonAMP_ =
+            moveLayoutRect(elementRect, opt_container.left, opt_container.top);
+      }
+      elementRectNorm = this.elementRectForNonAMP_;
+    }
+
+    // Normalize container element
+    const changeEntry = this.getValidIntersectionChangeEntry_(
+        elementRectNorm, viewport, opt_container);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   * @return {?IntersectionObserverEntry} A valid change entry.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, viewport, opt_container) {
+    // TODO: need to include owner LayoutRect info
+
+    // Building an IntersectionObserverEntry.
+    let intersectionRect = element;
+    if (opt_container) {
+      intersectionRect = rectIntersection(opt_container, element) ||
+          // No intersection.
+          layoutRectLtwh(0, 0, 0, 0);
+    }
+    intersectionRect = rectIntersection(viewport, intersectionRect) ||
+        // No intersection.
+        layoutRectLtwh(0, 0, 0, 0);
+
+    intersectionRect = moveLayoutRect(intersectionRect, -viewport.left,
+        -viewport.top);
+
+    const ratio = intersectionRatio(intersectionRect, element);
+    const newThresholdSlot = this.getThresholdSlot_(ratio);
+    if (newThresholdSlot == this.prevThresholdSlot_) {
+      return null;
+    }
+    this.prevThresholdSlot_ = newThresholdSlot;
+
+    // The element is relative to (0, 0), while the viewport moves. So, we must
+    // adjust.
+    const boundingClientRect = moveLayoutRect(element, -viewport.left,
+        -viewport.top);
+
+    // Now, move the viewport to (0, 0)
+    const rootBounds = moveLayoutRect(viewport, -viewport.left, -viewport.top);
+
+    return /** @type {!IntersectionObserverEntry} */ ({
+      time: performance.now(),
+      rootBounds: DomRectFromLayoutRect(rootBounds),
```

Comment:
> Or are you suggesting I change their param define to accept #DomRect as well?

They don't take a `LayoutRect` really, they take an object that has `left`, `right`, `top`, .... Since `DomRect`s have all that, they're already the right type. Since none of our code relies on `x` or `y`, we're good to just pretend they really are `LayoutRect`s.

> should we just consolidate the data type and add x y to LayoutRect?

Maybe.